src/sensors/lorreconstruction.rs

```diff
@@ -182,15 +184,21 @@ fn azimuthal_angle(hits: &[MCQT], xyzs: &SensorMap) -> Option<Vec<Angle>> {
 
 // Weighted mean and variance.
 // Might be better as a trait?
-// pub fn mean<T>(hits: &[T], weights: &[Ratio]) -> Option<T>
-// where
-//     T: Clone + std::iter::Sum + std::ops::Mul<Ratio> + std::ops::Div<Ratio>
-// {
-//     if hits.len() > 0 && hits.len() == weights.len() {
-//         let wsum: Ratio = weights.iter().cloned().sum();
-//         Some(hits.iter().cloned().zip(weights.iter().cloned()).map(|(h, w)| h * w / wsum).sum::<T>())
-//     } else { None }
-// }
+pub fn charge_weighted_mean<T>(data: &[T], weights: &[Ratio]) -> Option<T>
+where
+    T: Clone + std::iter::Sum + Mul<Ratio> + Div<Ratio>,
+    <T as Mul<Ratio>>::Output: Into<T>,
+    <T as Div<Ratio>>::Output: Into<T>
+{
+    if data.is_empty() { return None }
+    assert_eq!(data.len(), weights.len());
+    let sum_weights: Ratio = weights.iter().cloned().sum();
+    let weighted_sum: T = data.iter().cloned()
+        .zip(weights.iter().cloned())
+        .map(|(d, w)| (d * w).into())
+        .sum();
+    Some((weighted_sum / sum_weights).into())
+}
```

Comment:
We overlooked the (admittedly unlikely - but in Rust we tend to cater for rather than ignore the unlikely) possibility of `sum_weights` being zero, which would give a `NaN`. I suspect it should panic in this case. In fact, it should probably panic if *any* of the individual `weights` are negative.

Reply:
This still applies to the generic `weighted_mean` which replaces this function.

Reply:
The current version has an assert that checks all weights are positive (see here https://github.com/andLaing/petalo/blob/c2e5e41f728c81e5a3b7363d9abaa26bd19e9d47/src/sensors/lorreconstruction.rs#L275). So I think this can be resolved too.